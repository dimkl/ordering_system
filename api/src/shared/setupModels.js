const { Model } = require('objection');
const knex = require('./knex');

module.exports = () => { Model.knex(knex) }