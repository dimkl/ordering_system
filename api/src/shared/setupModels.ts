import { Model } from "objection";
import { knex } from "./knex";

export default () => {
  Model.knex(knex);
  return knex;
};
