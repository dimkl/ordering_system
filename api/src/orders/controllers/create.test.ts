/**
 * @integration-test true
 * @data-factory true
 */
import type { Knex } from "knex";

import setupModels from "../../shared/setupModels";

describe("POST /orders", () => {
  let knex: Knex;
  beforeAll(() => (knex = setupModels()));
  beforeEach(() =>
    knex.raw("truncate orders, order_items, customers, users, products, time_slots, slots cascade;")
  );
  afterAll(() => knex.destroy());

  it("creates and returns a order", async () => {
    const timeSlot = await DataFactory.createTimeSlot();

    const response = await request
      .post("/orders")
      .send({
        customer_id: timeSlot.customer.uuid,
        time_slot_id: timeSlot.uuid
      })
      .set("Accept", "application/json");

    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot({
      id: expect.any(Number),
      uuid: expect.any(String),
      created_at: expect.any(String),
      updated_at: expect.any(String),
      customer_id: expect.any(String),
      time_slot_id: expect.any(String)
    });
    expect(response.body.customer_id).toEqual(timeSlot.customer.uuid);
  });

  it("creates and returns bulk orders", async () => {
    const timeSlot = await DataFactory.createTimeSlot();
    const otherTimeSlot = await DataFactory.createTimeSlot({}, timeSlot.customer, timeSlot.slot);

    const data = [
      {
        customer_id: timeSlot.customer.uuid,
        time_slot_id: timeSlot.uuid
      },
      {
        customer_id: otherTimeSlot.customer.uuid,
        time_slot_id: otherTimeSlot.uuid
      }
    ];

    const response = await request.post("/orders").send(data).set("Accept", "application/json");

    expect(response.status).toBe(200);
    expect(response.body.length).toBe(2);
    expect(response.body[0]).toMatchSnapshot({
      id: expect.any(Number),
      uuid: expect.any(String),
      created_at: expect.any(String),
      updated_at: expect.any(String),
      customer_id: expect.any(String),
      time_slot_id: expect.any(String)
    });
    expect(response.body[1]).toMatchSnapshot({
      id: expect.any(Number),
      uuid: expect.any(String),
      created_at: expect.any(String),
      updated_at: expect.any(String),
      customer_id: expect.any(String),
      time_slot_id: expect.any(String)
    });

    expect(response.body[0].time_slot_id).toEqual(timeSlot.uuid);
    expect(response.body[1].time_slot_id).toEqual(otherTimeSlot.uuid);
  });

  it("creates and returns a order using internal customer_id", async () => {
    const timeSlot = await DataFactory.createTimeSlot();

    const response = await request
      .post("/orders")
      .send({ customer_id: timeSlot.customer.id, time_slot_id: timeSlot.id })
      .set("Accept", "application/json");

    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot({
      id: expect.any(Number),
      uuid: expect.any(String),
      created_at: expect.any(String),
      updated_at: expect.any(String),
      customer_id: expect.any(String),
      time_slot_id: expect.any(String)
    });
    expect(response.body.customer_id).toEqual(timeSlot.customer.uuid);
  });

  it("throws validation error for required properties", async () => {
    const response = await request.post("/orders").send({}).set("Accept", "application/json");

    expect(response.status).toBe(400);
    expect(response.body).toMatchSnapshot();
  });

  it("throws validation error for not existing customer_id", async () => {
    const timeSlot = await DataFactory.createTimeSlot();

    const response = await request
      .post("/orders")
      .send({
        customer_id: timeSlot.customer.uuid + "1",
        time_slot_id: timeSlot.uuid
      })
      .set("Accept", "application/json");

    expect(response.status).toBe(400);
    expect(response.body).toMatchSnapshot();
  });

  it("throws validation error for not existing time_slot_id", async () => {
    const timeSlot = await DataFactory.createTimeSlot();

    const response = await request
      .post("/orders")
      .send({
        customer_id: timeSlot.customer.uuid,
        time_slot_id: timeSlot.uuid + "1"
      })
      .set("Accept", "application/json");

    expect(response.status).toBe(400);
    expect(response.body).toMatchSnapshot();
  });

  it("omits additional properties", async () => {
    const timeSlot = await DataFactory.createTimeSlot();
    const customer = await DataFactory.createCustomer({
      email: "order.create@example.com"
    });

    const response = await request
      .post("/orders")
      .send({
        customer_id: customer.uuid,
        time_slot_id: timeSlot.uuid,
        created_at: "1680046371850"
      })
      .set("Accept", "application/json");

    expect(response.status).toBe(200);
    expect(response.body.created_at).not.toEqual("1680046371850");
    expect(response.body.customer_id).toEqual(customer.uuid);
  });
});
