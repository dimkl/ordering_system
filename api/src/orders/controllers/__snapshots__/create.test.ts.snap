// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /orders creates and returns a order 1`] = `
{
  "created_at": Any<String>,
  "customer_id": Any<String>,
  "id": Any<Number>,
  "state": "draft",
  "time_slot_id": Any<String>,
  "updated_at": Any<String>,
  "uuid": Any<String>,
}
`;

exports[`POST /orders creates and returns a order using internal customer_id 1`] = `
{
  "created_at": Any<String>,
  "customer_id": Any<String>,
  "id": Any<Number>,
  "state": "draft",
  "time_slot_id": Any<String>,
  "updated_at": Any<String>,
  "uuid": Any<String>,
}
`;

exports[`POST /orders creates and returns bulk orders 1`] = `
{
  "created_at": Any<String>,
  "customer_id": Any<String>,
  "id": Any<Number>,
  "state": "draft",
  "time_slot_id": Any<String>,
  "updated_at": Any<String>,
  "uuid": Any<String>,
}
`;

exports[`POST /orders creates and returns bulk orders 2`] = `
{
  "created_at": Any<String>,
  "customer_id": Any<String>,
  "id": Any<Number>,
  "state": "draft",
  "time_slot_id": Any<String>,
  "updated_at": Any<String>,
  "uuid": Any<String>,
}
`;

exports[`POST /orders throws validation error for not existing customer_id 1`] = `
[
  {
    "instancePath": "/customer_id",
    "keyword": "type",
    "message": "must be integer",
    "params": {
      "type": "integer",
    },
    "schemaPath": "defs#/definitions/id_or_uuid/anyOf/0/type",
  },
  {
    "instancePath": "/customer_id",
    "keyword": "format",
    "message": "must match format "uuid"",
    "params": {
      "format": "uuid",
    },
    "schemaPath": "defs#/definitions/id_or_uuid/anyOf/1/format",
  },
  {
    "instancePath": "/customer_id",
    "keyword": "anyOf",
    "message": "must match a schema in anyOf",
    "params": {},
    "schemaPath": "defs#/definitions/id_or_uuid/anyOf",
  },
]
`;

exports[`POST /orders throws validation error for not existing time_slot_id 1`] = `
[
  {
    "instancePath": "/time_slot_id",
    "keyword": "type",
    "message": "must be integer",
    "params": {
      "type": "integer",
    },
    "schemaPath": "defs#/definitions/id_or_uuid/anyOf/0/type",
  },
  {
    "instancePath": "/time_slot_id",
    "keyword": "format",
    "message": "must match format "uuid"",
    "params": {
      "format": "uuid",
    },
    "schemaPath": "defs#/definitions/id_or_uuid/anyOf/1/format",
  },
  {
    "instancePath": "/time_slot_id",
    "keyword": "anyOf",
    "message": "must match a schema in anyOf",
    "params": {},
    "schemaPath": "defs#/definitions/id_or_uuid/anyOf",
  },
]
`;

exports[`POST /orders throws validation error for required properties 1`] = `
[
  {
    "instancePath": "",
    "keyword": "required",
    "message": "must have required property 'customer_id'",
    "params": {
      "missingProperty": "customer_id",
    },
    "schemaPath": "#/required",
  },
  {
    "instancePath": "",
    "keyword": "required",
    "message": "must have required property 'time_slot_id'",
    "params": {
      "missingProperty": "time_slot_id",
    },
    "schemaPath": "#/required",
  },
]
`;
